src/MEstimators.jl

```diff
@@ -41,52 +42,52 @@ isconvex(::L2Estimator) = true
 The standard L1 M-estimator takes the absolute value of the residual, and is
 convex but non-smooth.
 """
-immutable L1Estimator <: MEstimator; end
+struct L1Estimator <: MEstimator; end
 L1Estimator(width) = L1Estimator()
-estimator_rho(r,::L1Estimator) = abs(r)
-estimator_psi(r,::L1Estimator) = sign(r)
-estimator_weight(r,::L1Estimator) = 1.0 ./ abs(r)
-estimator_sqrtweight(r,::L1Estimator) = abs(r) .^ (-0.5)
+estimator_rho(r,::L1Estimator) = abs.(r)
+estimator_psi(r,::L1Estimator) = sign.(r)
+estimator_weight(r,::L1Estimator) = 1.0 ./ abs.(r)
+estimator_sqrtweight(r,::L1Estimator) = abs.(r) .^ (-0.5)
 isconvex(::L1Estimator) = true
 
 """
 The convex L1-L2 estimator interpolates smoothly between L2 behaviour for small
 residuals and L1 for outliers.
 """
-immutable L1L2Estimator <: MEstimator; width::Float64; end
-estimator_rho(r,::L1L2Estimator) = 2.0*(sqrt(1.0 + 0.5*r.*r)-1.0)
-estimator_psi(r,::L1L2Estimator) = r ./ sqrt(1.0 + 0.5*r.*r)
-estimator_weight(r,::L1L2Estimator) = 1.0 ./ sqrt(1+0.5*r.*r)
-estimator_sqrtweight(r,::L1L2Estimator) = (1.0 + 0.5*r.*r) .^ (-1/4)
+struct L1L2Estimator <: MEstimator; width::Float64; end
+estimator_rho(r,::L1L2Estimator) = 2.0 * (sqrt.(1.0 .+ 0.5*r.*r) .- 1.0)
```

Comment:
In principle it may make more conceptual sense to use `.*` here rather than `*`.  `*` is a linear algebra operation that for vectors and scalars just happens to be the same as elementwise arithmetic. But what you're doing here is arguably elementwise arithmetic on residuals rather than linear algebra so the `.*` form is probably preferable.

Even better, you could just use `@.` to write all those dots for you:

```julia
@. 2.0 * (sqrt(1.0 + 0.5*r*r) - 1.0)
```

This observation about `*` vs `.*` goes for various operations further down as well.

Reply:
:+1:

Reply:
@c42f Quick question: how does that `@.` knows what operators to "lift"? Purely based on type matching (or mismatch in that case)?

Reply:
`@.` is a syntactic transformation which applies a `.` to every function/operator call. Then during code lowering any dotted function calls `f.(...)` are lifted to a call to `Base.broadcasted(f, ...)`. You can see this chain of events using some compiler introspection:

```julia
julia> @macroexpand @. 1 * b + c
:((+).((*).(1, b), c))

julia> Meta.@lower @. 1 * b + c
:($(Expr(:thunk, CodeInfo(
    @ none within `top-level scope'
1 ─ %1 = Base.broadcasted(*, 1, b)
│   %2 = Base.broadcasted(+, %1, c)
│   %3 = Base.materialize(%2)
└──      return %3
))))
```

Unfortunately the SSA form IR coming out of `Meta.@lower` doesn't look like normal julia code. But you can see that it's essentially `materialize(broadcasted(+, broadcasted(*, 1, b), c))`.

Reply:
So in short, the syntactic transformation is completely uniform; it doesn't know which parts to lift because it has no type info. So it just lifts all calls.

It's then up to the broadcast machinery (in particular the multiple dispatch in the `Base.broadcasted` implementation) to decide what to do based on the type of the inputs.

So having said all that, if some parts of your expression are purely scalar arithmetic it's arguably overkill to use `@.`. But I also think it's unlikely to cause trouble or inefficiency (but you can check).